File: plugin/src/main/java/com/oldstats/tracking/LootTracker.java
package com.oldstats.tracking;

import com.oldstats.api.OldStatsApiClient;
import com.oldstats.api.StatEvent;
import java.time.Duration;
import java.time.Instant;
import net.runelite.api.ItemComposition;
import net.runelite.api.NPC;
import net.runelite.client.events.NpcLootReceived;
import net.runelite.client.game.ItemManager;
import net.runelite.client.game.ItemStack;

/**
 * Tracks monster kills and the loot they drop via {@link NpcLootReceived},
 * which is posted by RuneLite's core LootManager service whenever a tracked
 * NPC's death is correlated with nearby item spawns — independent of
 * whether the built-in Loot Tracker plugin is enabled.
 */
public class LootTracker {
    private final OldStatsApiClient apiClient;
    private final ItemManager itemManager;

    private String lastBossKillName;
    private Instant lastBossKillAt;

    public LootTracker(OldStatsApiClient apiClient, ItemManager itemManager) {
        this.apiClient = apiClient;
        this.itemManager = itemManager;
    }

    public void onNpcLootReceived(NpcLootReceived event) {
        NPC npc = event.getNpc();
        String npcName = npc.getName() != null ? npc.getName() : "Unknown";
        boolean isBoss = BossList.isBoss(npcName);

        apiClient.enqueue(new StatEvent.Kill(npcName, npc.getId(), isBoss));

        if (isBoss) {
            lastBossKillName = npcName;
            lastBossKillAt = Instant.now();
        }

        for (ItemStack item : event.getItems()) {
            ItemComposition composition = itemManager.getItemComposition(item.getId());
            int price = itemManager.getItemPrice(item.getId());
            apiClient.enqueue(
                new StatEvent.Drop(npcName, composition.getName(), item.getId(), item.getQuantity(), (long) price * item.getQuantity())
            );
        }
    }

    /** Used by {@link PersonalBestTracker} to attribute a "new personal best" chat message to a boss. */
    public String recentBossKill(Duration within) {
        String name = lastBossKillName;
        Instant at = lastBossKillAt;
        if (name == null || at == null) return null;
        return Duration.between(at, Instant.now()).compareTo(within) <= 0 ? name : null;
    }
}

File: plugin/src/main/kotlin/com/oldstats/tracking/LootTracker.kt
package com.oldstats.tracking

import com.oldstats.api.OldStatsApiClient
import com.oldstats.api.StatEvent
import net.runelite.client.events.NpcLootReceived
import net.runelite.client.game.ItemManager
import java.time.Duration
import java.time.Instant

/**
 * Tracks monster kills and the loot they drop via [NpcLootReceived], which
 * is posted by RuneLite's core LootManager service whenever a tracked NPC's
 * death is correlated with nearby item spawns — independent of whether the
 * built-in Loot Tracker plugin is enabled.
 */
class LootTracker(
    private val apiClient: OldStatsApiClient,
    private val itemManager: ItemManager,
) {
    private var lastBossKillName: String? = null
    private var lastBossKillAt: Instant? = null

    fun onNpcLootReceived(event: NpcLootReceived) {
        val npc = event.npc
        val npcName = npc.name ?: "Unknown"
        val isBoss = BossList.isBoss(npcName)

        apiClient.enqueue(StatEvent.Kill(npcName = npcName, npcId = npc.id, isBoss = isBoss))

        if (isBoss) {
            lastBossKillName = npcName
            lastBossKillAt = Instant.now()
        }

        for (item in event.items) {
            val composition = itemManager.getItemComposition(item.id)
            val price = itemManager.getItemPrice(item.id)
            apiClient.enqueue(
                StatEvent.Drop(
                    npcName = npcName,
                    itemName = composition.name,
                    itemId = item.id,
                    quantity = item.quantity,
                    value = price.toLong() * item.quantity,
                )
            )
        }
    }

    /** Used by [PersonalBestTracker] to attribute a "new personal best" chat message to a boss. */
    fun recentBossKill(within: Duration): String? {
        val name = lastBossKillName ?: return null
        val at = lastBossKillAt ?: return null
        return if (Duration.between(at, Instant.now()) <= within) name else null
    }
}
